apps/small_filters/sobel.cpp

```diff
@@ -0,0 +1,61 @@
+#include "Halide.h"
+
+using namespace Halide;
+
+class Sobel : public Generator<Sobel> {
+public:
+    Input<Buffer<uint8_t>> input{"input", 2};
+    Output<Buffer<uint8_t>> output{"output", 2};
+    Var x{"x"}, y{"y"};
+    Func sobel_x_avg{"sobel_x_avg"}, sobel_y_avg{"sobel_y_avg"};
+    Func sobel_x{"sobel_x"}, sobel_y{"sobel_y"};
+
+    void generate() {
+        Func bounded_input{"bounded_input"};
+        bounded_input(x, y) = BoundaryConditions::repeat_edge(input)(x, y);
+
+        Func input_16{"input_16"};
+        input_16(x, y) = cast<uint16_t>(bounded_input(x, y));
+
+        sobel_x_avg(x,y) = input_16(x-1, y) + 2*input_16(x, y) + input_16(x+1, y);
+        sobel_x(x, y) = absd(sobel_x_avg(x, y-1), sobel_x_avg(x, y+1));
+
+        sobel_y_avg(x,y) = input_16(x, y-1) + 2*input_16(x, y) + input_16(x, y+1);
+        sobel_y(x, y) = absd(sobel_y_avg(x-1, y),  sobel_y_avg(x+1, y));
+
+        output(x, y) = cast<uint8_t>(clamp(sobel_x(x, y) + sobel_y(x, y), 0, 255));
```

Comment:
Should probably comment here that this isn't the standard sobel filter (which uses the magnitude of the gradient), this doesn't need a square root.